src/ncp/spinel.c

```diff
@@ -332,7 +331,14 @@ spinel_datatype_vunpack_(const uint8_t *data_ptr, spinel_size_t data_len, const
 
             if (arg_ptr)
             {
-                *arg_ptr = (spinel_ipv6addr_t *)data_ptr;
+                if (*arg_ptr)
+                {
+                    memcpy(*arg_ptr, data_ptr, sizeof(spinel_ipv6addr_t));
+                }
+                else
+                {
+                    *arg_ptr = (spinel_ipv6addr_t *)data_ptr;
+                }
```

Comment:
I fully agree that it's a good practice to initialize all local variables (including pointers).

However, in the case of the use pattern for `spinel_datapack_unpack()` the fact that a pointer variable may be un-initialized is not really unsafe or error-prone:

- What we are sending to `unpack()` function is actually a pointer to a variable and we expect the `unpack()` function to update that variable. Basically the variable is used as an output of the `unpack()` function. For some simpler data types like `UINT16` the variable is a `uint16_t` but for more complex data types the variable happens to be a pointer to something. 

- If the `unpack()` is successful, the variable is updated and should then contain a valid value, so when the variable itself is a pointer, if the unpacking is successful it should be updated to point to what was unpacked. It should not matter if a pointer variable was initialized or not (the same way that it should not matter if the `uint16_t` variable was initialized). The key is to check for success of `unpack()` operation before using any of the output variables (including pointer types).

Your change in this PR makes the `unpack()` treat the case where the variable associated with the type is itself a pointer to something, differently than other simpler types. If the type is a pointer to something, with your change whether the variable is `NULL` or not, matters and changes the behavior (if it is `NULL` we update the variable, otherwise we copy some content at the address being pointed by variable and not change the variable itself).

The impact of this change on other projects using `spinel.c` is not limited to ensuring that pointer variables are initialized to `NULL` when they are declared.  For example, consider the code performing the unpacking of some table (say IPv6 address table, on-mesh prefixes, child table, whitelist, ...).  As we loop through and unpack each entry in the table, we typically reuse the same set of variables. With your changes to keep the behavior as before, we need to ensure to set all pointer type variables to `NULL` every time before calling the `unpack` (or any of its many derivatives). This requires some careful changes to all other projects and honestly I think it is something that can be easily missed now or used incorrectly in future. That's why I think the risks of this change far outweigh any possible benefit/flexibility it may provide.